### ISSUE TITLE
[PyTreeArray] fix multiplication of pytrees

### ISSUE THREAD

**Issue by PhilipVinc** [PhilipVinc/netket_dynamics#2]
@inailuig probably you should move https://github.com/inailuig/pytreearray/commits/nested here?

**Comment by inailuig** [comment]
You are referring to the last 2 commits there which are supposed to fix 1d@1d multiplication?
iirc the second one just reverts the 1st one

Numpy returns a np.float64 object for 1d@1d, jnp returns a Devicearray with empty shape.

Maybe its easier to just return the `tree` (like np) and not a `PyTreeArray(tree, (), ())` (like jnp).


@PhilipVinc Any other issues?

___

For the reference:

the idea of that whole branch (i.e. this commit: https://github.com/inailuig/pytreearray/commit/f5fa65995c9bd47dabf2622c9be45bddcdcdb434) is the following:

instead of storing a tuple of treedefs and axes on the top level, you have nested pytreearrays where the tuple of treedefs and axes is constructed recursively from the children.

This makes the code more convoluted (and probably less efficient; compile time wise) but has the advantage that one can do the following and obtain the correct pytree matrix at the end:
```python
import jax
import jax.numpy as jnp
import pytreearray as pta
from functools import partial

def f(p, x):
     return jax.lax.sin(x@p['a'])*p['b']

N=10
d = 5

x = jnp.ones((N,d))
p = pta.PyTreeArray1({'a': jnp.ones(d), 'b':jnp.ones(1)})

def hessian(f):
     return jax.jacfwd(jax.jacrev(f))


h = hessian(partial(f,x=x))(p)

print('treedefs', h.treedefs)
print(h@p)
``` 
```
treedefs: (PyTreeDef({'a': *, 'b': *}), PyTreeDef({'a': *, 'b': *}))

PyTreeArray(tree_={'a': DeviceArray([[5.078284, ....})
```

**Comment by inailuig** [comment]
@PhilipVinc 
Actually there is a way to make it work for jacrev:

```python
def f2(*args, **kwargs):
    return pta.PyTreeArray1(f(*args, **kwargs))

r = jax.jacrev(partial(f,x=x))(p)
print(r.treedefs, r.axes)
```
```
(PyTreeDef(*), PyTreeDef({'a': *, 'b': *})) (1, {'a': 1, 'b': 1})
```
